File: ios/Plugin/ExternalLinkAccountPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(ExternalLinkAccountPlugin, "ExternalLinkAccount",
           CAP_PLUGIN_METHOD(canOpen, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(open, CAPPluginReturnPromise);
)

File: ios/Plugin/ExternalLinkAccountPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ExternalLinkAccountPlugin)
public class ExternalLinkAccountPlugin: CAPPlugin {
    private let implementation = ExternalLinkAccount()

    @objc func canOpen(_ call: CAPPluginCall) async {
        call.resolve([
            "value": await implementation.canOpen()
        ])
    }

    @objc func open(_ call: CAPPluginCall) async {
        do {
            try await implementation.open()
            call.resolve()
        } catch ExternalLinkAccountError.implementManually {
            call.reject("ExternalLinkAccount modal must be implemented manually", nil, ExternalLinkAccountError.implementManually)
        } catch ExternalLinkAccountError.cannotOpen {
            call.reject("ExternalLinkAccount modal cannot be open", nil, ExternalLinkAccountError.cannotOpen)
        } catch {
            call.reject("ExternalLinkAccount modal cannot be open", nil, error)
        }
    }
}
